SJRK-405: Implement server side logic for signup/login/logout and session support

jobara opened the pull request:
https://issues.fluidproject.org/browse/SJRK-405

jobara committed:
SJRK-405: (WIP) exploring user account system

jobara committed:
SJRK-405: (WIP) basic signup, login and logout endpoints

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Move request handlers to requestHandlers.js

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Add status codes for errors

jobara committed:
SJRK-405: Add validation of request parameters

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Update to latest dev release of fluid-express-user

jobara committed:
SJRK-405: Add authentication requirements for saving

- Also linting

jobara committed:
SJRK-405: Update to latest dev release of fluid-express-user

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Add view for fetching story for edit.

Restricted by authenticated user.

jobara committed:
SJRK-405: Fix broken tests

jobara committed:
SJRK-405: Signup and Login with email address

jobara committed:
NOJIRA: Update dependencies

jobara committed:
SJRK-417: Update linting config and lint files

- Includes sourcing global declarations for fluid and sjrk from the
eslint config.

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Remove test session endpoint

jobara committed:
SJRK-405: Add dataSource tests

jobara committed:
SJRK-405: Clean up comments for dataSource-tests

BlueSlug commented on src/server/dataSource.js:
```diff
@@ -64,6 +64,22 @@ fluid.defaults("sjrk.storyTelling.server.dataSource.couch.view", {
+// A CouchDB DataSource for retrieving a story by authorID and storyId
+fluid.defaults("sjrk.storyTelling.server.dataSource.couch.authorStoriesView", {
+    gradeNames: ["sjrk.storyTelling.server.dataSource.couch.core"],
+    // TODO: investigate overriding the resolveUrl invoker to all for providing the query paramters via
```
Could you file a Jira for this and link to it from this comment, please?

jobara replied:
https://issues.fluidproject.org/browse/SJRK-443

jobara committed:
SJRK-405: Linting

jobara committed:
SJRK-405: Exclude couchViews from NYC instrumentation.

NYC was instrumenting the couchdb views causing the datasource tests to fail because the NYC injected code couldn't be found when running the view in pouchDB.

jobara committed:
SJRK-405: Update to latest stable release of fluid-express-user

BlueSlug commented on src/server/requestHandlers.js:
```diff
@@ -131,15 +132,61 @@ sjrk.storyTelling.server.handleGetStory = function (request, dataSource) {
     });
 };
 
+// Kettle request handler for a single story (View)
```
```suggestion
// Kettle request handler for getting a single story (Edit)
```

BlueSlug commented on src/server/requestHandlers.js:
```diff
@@ -131,15 +132,61 @@ sjrk.storyTelling.server.handleGetStory = function (request, dataSource) {
+});
+
+/**
+ * Gets a single story's data from the database, parses it, rebuilds any URLs to
```
No URL rebuilding is happening any more as of `0.4.0`, I forgot to remove that part of the comment for `sjrk.storyTelling.server.handleGetStory`. This part of the sentence should be removed for that function as well as this one here

BlueSlug commented on src/server/requestHandlers.js:
```diff
@@ -583,3 +660,143 @@ fluid.defaults("sjrk.storyTelling.server.nodeModulesHandler", {
+
+/********************************************
+ * Request Handlers for signup/login/logout *
+ *******************************************/
```
This comment style (along with its partner at the bottom) doesn't match the rest of the project. I see that it's being used to delineate which grades are handlers for stories & files and which are for session management, and I wonder if it might not make more sense to split these into their own file, e.g. `requestHandlers-session.js` and `requestHandlers-story.js`, or some other more meaningful/practical naming convention. What do you think?

BlueSlug commented on src/server/serverSetup.js:
```diff
@@ -41,15 +45,65 @@ fluid.defaults("sjrk.storyTelling.server", {
+                    },
+                    couchDBAuthorsURL: {
+                        record: "@expand:kettle.resolvers.env(COUCHDB_URL)",
+                        target: "{that gexpressUserUtils}.options.dataSourceConfig.host"
```
```suggestion
                        target: "{that expressUserUtils}.options.dataSourceConfig.host"
```

BlueSlug commented on src/server/requestHandlers.js:
```diff
@@ -583,3 +660,143 @@ fluid.defaults("sjrk.storyTelling.server.nodeModulesHandler", {
+ * Logs in an existing user with the password and username provided in the request.
+ *
+ * @param {Object} request - a Kettle request
+ * @param {fluid.express.user.utils} expressUserUtils - an instance of `fluid.express.user.utils`
```
In other parts of the codebase, I've put `{Component}` in cases where the arg is an Infusion component. For consistency, I'd suggest either setting this to that or setting all the other ones to the fully-qualified grade name. The latter might be out of scope for this Jira, in which case we should file a new one to look into what we should be doing and then implementing whatever solution we come up with.

jobara replied:
I've switched it to `{Component}` for consistency. We should review this in the future. In Infusion the practice has moved to use the component grade; however, a description is still required by the JSDoc validation.

BlueSlug commented on src/server/serverSetup.js:
```diff
@@ -41,15 +45,65 @@ fluid.defaults("sjrk.storyTelling.server", {
+                    expressUserUtils: {
+                        type: "fluid.express.user.utils",
+                        options: {
+                            iterations: 10,
```
Is 10 iterations enough? I've seen 100, 1,000 or even 10,000 iterations suggested in order to defend against brute-force attacks, especially so with faster hashing algorithms. That said, if we're heading towards cloud hosting, too high a number could make hosting very expensive...

jobara replied:
This is the default from the underlying library actually. We could make it some other value as needed.

BlueSlug commented on src/server/serverSetup.js:
```diff
@@ -252,3 +334,11 @@ sjrk.storyTelling.server.getThemePath = function (theme, themeFolder) {
 };
+
+sjrk.storyTelling.server.makeMemorySessionStore = function () {
+    // TODO: Currently using https://www.npmjs.com/package/memorystore as it is a production ready memory store;
```
If this `TODO` is not intended to be tackled in SJRK-405, please file a Jira to do that work and add a link to that Jira here

jobara replied:
https://issues.fluidproject.org/browse/SJRK-444

BlueSlug commented on src/server/serverSetup.js:
```diff
@@ -252,3 +334,11 @@ sjrk.storyTelling.server.getThemePath = function (theme, themeFolder) {
     return themePath;
 };
+
+sjrk.storyTelling.server.makeMemorySessionStore = function () {
```
Please add a JSDoc comment for this function

BlueSlug commented on tests/server/dataSource-tests.js:
```diff
@@ -0,0 +1,606 @@
+        }
+    },
+    stories: {
+        "story1": {
```
It may be helpful to name these test cases after their contents or purpose

BlueSlug commented on tests/server/dataSource-tests.js:
```diff
@@ -0,0 +1,606 @@
+ * @param {Object} expected - The expected value to compare against the `actual` value.
+ * @param {Object} actual - The actual value, minus `_rev` property, compared against the `expected` value.
+ */
+sjrk.test.storyTelling.server.dataSource.tester.assertResponse = function (msg, expected, actual) {
```
A similar filtering out of `_rev` is done in `sjrk.storyTelling.server.testServerWithStorageDefs.verifyStoryPersistence`. Perhaps that piece of code and this function could be combined in a shared function in `serverTestUtils.js`.

BlueSlug commented on tests/server/dataSource-tests.js:
```diff
@@ -0,0 +1,606 @@
+ */
+sjrk.test.storyTelling.server.dataSource.tester.assertMissingError = function (msg, actual, options) {
+    var tokens = fluid.merge("replace", {}, {method: "GET", storyID: ""}, options || {});
+    var expected = {
```
Would the expected values be better defined outside of this function and passed in?

BlueSlug reviewed (commented):
Looking pretty good!

jobara committed:
SJKR-405: Remove username from author doc

jobara committed:
SJRK-405: Update comments

Co-authored-by: BlueSlug <mossmonkey@gmail.com>

jobara committed:
SJRK-405: Fix typo

Co-authored-by: BlueSlug <mossmonkey@gmail.com>

jobara committed:
SJRK-405: Cleaning up comments

jobara committed:
SJRK-405: Update JSDoc

jobara committed:
SJRK-405: Clean up tests based on code review.

jobara committed:
SJRK-405: Group request handlers into separate files

jobara committed:
SJRK-405: Fix issue where signup doesn't recognize existing account.

jobara committed:
SJRK-405: Return email for logged in user on

jobara committed:
SJRK-405: Add expressUserUtils config to secrets file

jobara committed:
SJRK-405: Moved defaults for expressUserUtils back to component tree.

Overrides can still be provided via the secrets.json file

jobara committed:
SJRK-405: Refactor testDBs to shared mockDatabase grade

jobara committed:
SJRK-405: replace fluid.merge with fluid.extend

BlueSlug commented on tests/server/dataSource-tests.js:
```diff
@@ -0,0 +1,565 @@
+        viewDataSource: {
+            type: "sjrk.storyTelling.server.dataSource.couch.view",
+            options: {
+                host: "http://localhost:6789"
```
Perhaps a distributeOptions could be used to inject this host configuration into each test dataSource. This is fine either way, though :)

BlueSlug commented on tests/server/dataSource-tests.js:
```diff
@@ -0,0 +1,565 @@
+    },
+    testOpts: {
+        storiesViewError: {
+            "reason": "ddoc _design/stories has no view named notAView",
```
Is this meant to read "ddoc" or "doc"? If it's the former, it may be more clear if it says "design doc"

jobara replied:
that's an error straight from the database, it isn't something I defined.

BlueSlug replied:
Curious! But okay, let's just leave it as-is, in that case

jobara committed:
SJRK-405: Add tests for signup/login/logout

jobara committed:
SJRK-405: Update comment about ajax failure cases

cindyli commented on src/server/db/authors-dbConfiguration.js:
```diff
@@ -0,0 +1,50 @@
+fluid.defaults("sjrk.storyTelling.server.authorsDb", {
+    gradeNames: ["fluid.couchConfig.pipeline.retrying"],
+    couchOptions: {
+        dbName: "authors"
```
Just curious, is "authors" a separated database that only contains the "user" document type? What's the reason to decide on using a separate database rather than making use of the existing "stories" database with a new "user" document type?

With the current logic, using 2 databases works the same way. But it feels strange using multiple databases to support an app, which is unnecessary.

jobara replied:
That the user accounts are managed through fluid-express-users library, but I think it could actually be in the same database as stories. I'll look into that.

BlueSlug replied:
I looked into this when I first saw @jobara's choice. Turns out using separate databases is actually preferable, on account of the way index building works in CouchDB, though the performance gains to us will be minimal. Here's a very interesting thread on the topic, I recommend reading the answers other than the one the OP chose: https://stackoverflow.com/questions/10503089/couchdb-multiple-databases

jobara replied:
@BlueSlug that's a very interesting point. Unfortunately for me I already combined them. @cindyli and @BlueSlug let me know if you think we should split them back out.

cindyli replied:
It's great to know separating data into multiple databases helps with the CouchDB performance. Thanks for the good thread, @BlueSlug. After reading the thread, I'm fine with either.

If @jobara has already combined it, I would prefer leave as combined. As @BlueSlug said, the performance gain for SJRK will be minimal. Moreover, using one database simplifies the code and is easier for backup and maintenance.

cindyli commented on src/server/db/singleNodeConfiguration.js:
```diff
@@ -44,3 +44,11 @@ fluid.defaults("sjrk.storyTelling.server.globalChangesDb", {
 });
+
+// includes the users database name
+fluid.defaults("sjrk.storyTelling.server.storyUsersDb", {
```
What is this "users" database? It doesn't seem to be used anywhere.

jobara replied:
I believe it's couchDB's user database for accounts to itself.

jobara replied:
oh actually this one might be a mistake, might have been leftovers from trying things out early on.

jobara committed:
SJRK-405: remove unused  users database

jobara committed:
SJRK-405: Fix  sourcing ID from provided story in save request.

cindyli commented on src/server/storyRequestHandlers.js:
```diff
@@ -157,28 +202,51 @@ fluid.defaults("sjrk.storyTelling.server.saveStoryHandler", {
+
+    dataSource.get({directStoryId: id}).then(function (response) {
+        if (request.req.session.authorID === response.authorID) {
+            sjrk.storyTelling.server.saveStoryToDatabase(dataSource, request.req.session.authorID, request.req.body, request.events.onSuccess, request.events.onError);
```
Could you add a code comment here and at line 226 that this line is to update an existing story while line 226 is to create a new story.

cindyli commented:
It would be helpful to add server API documentation when you have time. Thanks.

jobara committed:
SJRK-405: Combine authors and storeis db

jobara committed:
SJRK-405: Simplify auth endpoints

jobara committed:
SJRK-405: Rename auth tests

jobara committed:
SJRK-405: Return authorID in story dataSource

jobara committed:
SJRK-405: fix handGetStory response

jobara committed:
SJRK-405: Prevent updating published stories created by a guest

jobara committed:
SJRK-405: Fix error status code for getStory

jobara committed:
SJRK-405: Add/Update server tests to cover authentication

jobara committed:
Merge branch 'main' into SJRK-405

jobara committed:
SJRK-405: Lint test files

Remove unused variables and includes.

jobara committed:
SJRK-405: Fix tests to work with windows

On windows different "errno" are reported, so just checking the error "code".

jobara committed:
SJRK-405: Add new handlers and datasources to docs

jobara committed:
SJRK-405: Add authentication handlers to server docs

jobara committed:
SJRK-405: Make checkPeriod configurable

jobara commented:
@cindyli this PR is ready for review.

cindyli commented on src/server/db/story-dbConfiguration.js:
```diff
@@ -141,7 +151,7 @@ sjrk.storyTelling.server.storiesDb.storyTagsFunction = function (doc) {
  */
 sjrk.storyTelling.server.storiesDb.storiesByIdFunction = function (doc) {
-    if (doc.value.published) {
+    if (doc.type === "story" && doc.value.published) {
```
The condition of checking `doc.value.published` shows a more meaning name for this view might be "publishedStoriesById`.

jobara replied:
Makes sense. I've renamed it.

jobara committed:
SJRK-405: Rename stories view

Based on PR feedback it was suggested that a more meaningful name would be "publishedStoriesById".

cindyli commented on src/server/serverSetup.js:
```diff
@@ -41,15 +46,69 @@ fluid.defaults("sjrk.storyTelling.server", {
+                        secret: "{server}.options.secureConfig.secrets.session"
+                    }
+                },
+                account: "{that}.options.secureConfig.secrets.account",
```
What is this account for? `secrets.json.example` does not provide an example of this account value.

This affects line 69-72 too which distributes this value.

jobara replied:
This provides a means of configuring [`fluid.express.user.utils`](https://github.com/fluid-project/fluid-express-user/blob/e826290beb4fdb416a4fedf602803ac1bcef4f37/src/js/server/utils.js#L28). Not sure how to show that in the `secrets.json.example`. Any suggestions?

cindyli replied:
Can you point me to the doc or code that show how this `account` is used for configuring `fluid.express.user.utils`? I didn't find its usage in the given link. Just want to get an idea about its value type and the logic behind it. Thanks.

jobara replied:
@cindyli the link above is to the components defaults. We add fluid.express.user.utils as a [subcomponent](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L76-L102) to the server. These defaults can be set/overridden by passing in configuration to the secrets.json file under the `account` property. These are relayed via the lines you mentioned above. The component options for [account](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L55) and the [options distributions](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L69-L72) to send this to the instance of fluid.express.user.utils.

cindyli replied:
I think now I understand better of what you said about "provides a means of configuring fluid.express.user.utils". A couple of suggestions:

1. Please consider to use a more meaningful variable name such as `expressUserUtilsConfig`;
2. I'm not sure if it's a good idea to use `secrets.json` to distribute component options. These options only make sense to SJRK developers while `secrets.json` gives a feeling of being configured by website admins who unlikely know about infusion and what component options are;
3. If you decided to use `secrets.json` for this config, please add documentation, for example in README where `secrets.json` is explained. (If `secrets.json` is a json5 file, comments for each value in that file would be easier to access.)

jobara replied:
@cindyli I've added some documentation about configuration, add some examples to the secrets.json.example file, and renamed some things. I think mostly you can just use the kettle config to configure expressUserUtils, but there are some parts that we may want to express in the secrets file, so I've illustrated those in the example.

cindyli commented on src/server/serverSetup.js:
```diff
@@ -207,6 +279,21 @@ fluid.defaults("sjrk.storyTelling.server.app.storyTellingHandlers", {
             route: "/clientConfig",
             method: "get"
         },
+        signupHandler: {
```
This pull request does not contain the tests and documentation for these new endpoints. Is it planned to be addressed in a separate ticket?

jobara replied:
The authentication related handlers are included in [testSeverWithStorage.js](https://github.com/jobara/sjrk-story-telling/blob/SJRK-405/tests/server/testServerWithStorage.js) as they are required for saving/editing stories.

jobara replied:
They are also tested in [/testServerAuthentication.js](https://github.com/jobara/sjrk-story-telling/blob/SJRK-405/tests/server/testServerAuthentication.js)

cindyli commented on src/server/db/story-dbConfiguration.js:
```diff
@@ -170,3 +191,16 @@ sjrk.storyTelling.server.storiesDb.validateFunction = function (newDoc) {
+ */
+sjrk.storyTelling.server.storiesDb.byUsernameOrEmail = function (doc) {
+    if (doc.type === "user") {
+        emit(doc.username, doc);
```
I know nothing about the client side logic. As the server logic in this pull request always returns the email address as a login success signal, I wonder if a user could login via either the username or email address, or just email address. In the case of the latter, it might be better to remove this line in order to reduce the index volume and prevent a potential security issue.

jobara replied:
There are no explicit usernames anymore, only an e-mail address to identify authors. Thanks for spotting this leftover bit. I've removed it, and have also renamed the name of the function. There are some other references to byUsernameOrEmail but these are used internally by fluid-express-user, so I have left them as is.

jobara committed:
SJRK-405: Cleanup view for finding users as only e-mail addresses are used now.

jobara committed:
SJRK-405: Add session endpoint

Can be used to check if a session is stil active.

jobara commented:
@cindyli I'm trying to write up some docs on the server endpoints as well. Working through how to format them, but I have an idea now.

jobara committed:
SJRK-405: Add documentation for server endpoints

jobara commented:
@cindyli I've added the server endpoints documentation. Ready for more review.

cindyli commented:
The server endpoints doc looks great. Thanks, @jobara.

jobara committed:
SJRK-405: Cleanup configuration for express user utils.

jobara committed:
SJRK-405: Linting

jobara committed:
SJRK-405: Update code highlighting for comments

cindyli commented on src/server/serverSetup.js:
```diff
@@ -41,15 +46,69 @@ fluid.defaults("sjrk.storyTelling.server", {
+                        target: "{that > kettle.middlewareHolder > session}.options"
+                    },
+                    authorCredentialConfig: {
+                        source: "{that}.options.account",
```
The differentiation of the kettle server config and account credential config in various config files makes a lot of sense. Thanks, @jobara.

With the renaming at line 55, the source reference of `account` should also be renamed to `authorCredentialConfig`. It would be nice to have tests for the proper handling/distribution of secrets.json.

jobara replied:
Thanks for catching the leftover bit. I've addressed that and added some test for the distributions of the secrets.

jobara committed:
SJRK-405: Correct source for authorCredentialConfig distribution

jobara committed:
SJRK-405: Added tests for secrets distributions

cindyli commented:
This pull request looks great to me. 

@BlueSlug, can you help to do a full manual test? Once it passes the test, feel free to merge. Thanks.

jobara committed:
SJRK-405: Remove secrets.json in tests from gitignore

BlueSlug commented on docs/CONFIGURATION.md:
````diff
@@ -0,0 +1,75 @@
+
+```JSON5
+{
+    "adminPass": "admin password",
````
The section of `README.md` titled "Running the site" mentions what the contents of `secrets.json` need to be in order to run the site. I suggest either updating that section to reflect the contents here, or linking directly to this file

BlueSlug reviewed (approved):
The tests are all passing without issue, now :)

jobara committed:
SJRK-405: Update README to point at CONFIGURATION docs

jobara commented:
I believe @BlueSlug gave his [approval](https://matrix.to/#/!mDigWuPWilYGiEcOCA:matrix.org/$8lcQ77Vz-nE2jIlbd6AQEq1lv3l7XjkG163ta1EeYUg?via=matrix.org) in the channel. I'll self merge this PR.

jobara commented:
Merged at 3845afd05c22ba96283c4558581b90e69ea47319
